Add a 'timer' function

ericapisani opened the pull request:
The 'timer function' can be used to record the amount of time a function takes, and then send that value using the 'timing' function.

ericapisani committed:
Add a 'timer' function that can be used to record the amount of time a function takes, and then send that value using the 'timing' function.

ericapisani committed:
Add unit test to check that additional arguments (such as tags) are passed correctly to the tming function.

ericapisani committed:
Add example of timer function to README.

ravangen commented on lib/statsd.js:
```diff
@@ -107,6 +107,34 @@ Client.prototype.timing = function (stat, time, sampleRate, tags, callback) {
+ * Function that records the time to run a given function
+ * and sends the recorded time (in milliseconds)
+ * @param func {Function} The function to run
+ * @param metricsArgs {Object} The arguments to be used by Client.prototype.timing
```
Can you switch this to use the flat list of parameters so this matches other function interfaces. Where we don't need to provide `time {Number}`. I do wonder if `func` should be the first or second parameter since all the other functions have the `stat` name as the first one I think.

```javascript
  * @param stat {String|Array} The stat(s) to send
  * @param sampleRate {Number=} The Number of times to sample (0 to 1). Optional.
  * @param tags {Array=} The Array of tags to add to metrics. Optional.
  * @param callback {Function=} Callback when message is done being delivered. Optional.
```

ericapisani replied:
Sure thing 👍  

However, if this uses flat list of parameters rather than an object, then it's probably best to have the `func` as the first parameter so that the metrics args are at least grouped together in the order that they usually appear in other functions, especially since `func` is a non-optional argument here (whereas some of the metrics args are)

ravangen commented on README.md:
```diff
@@ -88,6 +88,12 @@ The check method has the following API:
+  // Timer: Records how long a given function takes to execute (in milliseconds),
+  // and then sends that value using 'client.timing'
+  var fn = function(a, b) { return a + b };
+  var metricsArgs = { 'stat': 'response_time' };
```
`my_timer` perhaps to match other examples since `fn` isn't really a `response_time`

ericapisani replied:
The reasoning behind this was to match the above `timing` example, which uses `response_time` as their `stat` value, but I'd be more than happy to change this in order to convey that this is a 'time to execute' stat.

Would `fn_execute_time`/`execute_time` be a more meaningful name? Or maybe something else?

ravangen replied:
`response_time` works I think you can leave it as you have it. Originally I was thinking about an api response time instead of a function time.

ravangen commented on lib/statsd.js:
```diff
@@ -107,6 +107,34 @@ Client.prototype.timing = function (stat, time, sampleRate, tags, callback) {
 };
 
+/**
+ * Function that records the time to run a given function
```
Perhaps something like: `Represents the timing stat by recording the duration a function takes to run (in milliseconds)`

This way the description indicates what metrics will be output, similar to other descriptions

ericapisani commented on test/test_statsd.js:
```diff
@@ -487,6 +487,55 @@ function doTests(StatsD) {
+    it('should send stat and time to execute to timing function', function(finished) {
+      udpTest(function (message, server) {
+        // Search for a string similar to 'test:0.123|ms'
+        var re = RegExp("(test:)([0-9]+\.[0-9]+)\\|{1}(ms)");
```
regex has to be used because the `value` sent will always change (since it's the time taken to execute the function`

ericapisani committed:
- Update the description of the 'timer' function
- Flatten list of metrics parameters going to the 'timer' function so it's more like the other functions in the library.
- Update README example with more accurate stat name.

ravangen commented:
+1

ericapisani committed:
Use .apply in order to properly pass the values in the arguments object to the user-provided function.

ravangen commented on lib/statsd.js:
```diff
@@ -107,6 +107,36 @@ Client.prototype.timing = function (stat, time, sampleRate, tags, callback) {
+  return function () {
+    var start = process.hrtime();
+    try {
+      return func.apply(null, arguments);
```
Is `null` adequate or do should we pass through `this` value for `func`?

ericapisani committed:
Update README description to be more explicit about what 'timer' returns and what the parameters are.

ericapisani committed:
Fix accidental typo in describing which client function the parameters after 'fn' belong to.

ericapisani commented:
These changes have been merged into master in the upstream repo (brightcove). Closing.
